src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -119,12 +144,35 @@ async Task EnsureLong()
 
                 try
                 {
+                    var timeoutCt = CancellationToken.None;
+                    if (async)
+                    {
+                        if (Timeout > 0)
+                        {
+                            // We reuse the timeout's cancellation token source as long as it hasn't fired, but once it has
+                            // there's no way to reset it (see https://github.com/dotnet/runtime/issues/4694)
+                            var timeoutTimeSpan = TimeSpan.FromMilliseconds(Timeout);
+                            _timeoutCts.CancelAfter(timeoutTimeSpan);
+                            if (_timeoutCts.IsCancellationRequested)
+                            {
+                                _timeoutCts.Dispose();
+                                _timeoutCts = new CancellationTokenSource(timeoutTimeSpan);
+                            }
+                            timeoutCt = _timeoutCts.Token;
+                        }
+                    }
+                    else
+                    {
+                        if (Timeout != _currentTimeout)
+                            Underlying.ReadTimeout = _currentTimeout = Timeout;
+                    }
+
                     var totalRead = 0;
                     while (count > 0)
                     {
                         var toRead = Size - FilledBytes;
                         var read = async
-                            ? await Underlying.ReadAsync(Buffer, FilledBytes, toRead)
+                            ? await Underlying.ReadAsync(Buffer, FilledBytes, toRead, timeoutCt)
```

Comment:
The same cancellation token is used across multiple reads - this means that the timeout is aggregated. This is unlike for the sync timeouts, where each ReadAsync will have its own timeout.

We shouldn't have this kind of discrepancy between the sync and async timeouts - can you please reset the cancellation token source for each read in the loop?

After this, it probably makes sense to have two separate sync/async implementations, rather than the current code where sync/async is checked three different times.

Reply:
I completely agree. The timeout should be per `ReadAsync ()` and not span multiple of those for consistency reasons.
This also means that higher lever APIs that need some higher level timeout (time between two complete messages not only reads of any sort of data from the socket) will need to roll this on their own, but this is probably something that we can't prevent anyways since a timeout spanning multiple reads wouldn't necessarily mean that a complete message got read.